cognite/powerops/resync/v2/shop_to_assets.py

```diff
@@ -0,0 +1,375 @@
+
+        return assets_by_xid
+
+    def _to_generator(self, name: str, data: dict) -> GeneratorWrite:
+        curve = data["gen_eff_curve"]
+        efficiency_curve = GeneratorEfficiencyCurveWrite(
+            external_id=f"{name}_efficiency_curve", ref=curve["ref"], power=curve["x"], efficiency=curve["y"]
+        )
+        turbine_curves = []
+        for curve in data["turb_eff_curves"]:
+            turbine_curve = TurbineEfficiencyCurveWrite(
+                external_id=f"{name}_turbine_eff_curve_ref_{curve['ref']}",
+                head=curve["ref"],
+                flow=curve["x"],
+                efficiency=curve["y"],
+            )
+            turbine_curves.append(turbine_curve)
+
+        # Start cost is assumed to be a dictionary with (timestamp, value) key-value pairs
+        # We assume it is constant and thus use only the first value.
+        startcost = next(iter(data["startcost"].values()))
+
+        generator_timeseries = self.times_series_by_generator_name.get(name, {})
+
+        return GeneratorWrite(
+            external_id=f"generator_{name}",
+            name=name,
+            p_min=data["p_min"],
+            penstock=data["penstock"],
+            start_cost=startcost,
+            start_stop_cost=generator_timeseries.get("start_stop_cost"),
+            is_available_time_series=generator_timeseries.get("is_available"),
+            efficiency_curve=efficiency_curve,
+            turbine_curves=turbine_curves,
+        )
+
+    def _to_plant(
+        self,
+        name: str,
+        data: dict,
+        plant_display_name_and_order: dict,
+        watercourse_xid: str,
+        reservoir_by_name: dict[str, ReservoirWrite],
+        generator_by_name: dict[str, GeneratorWrite],
+        shop_model: dict,
+    ) -> PlantWrite:
+        plant_timeseries = self.times_series_by_plant_name.get(name, {})
+        display_name, order = plant_display_name_and_order.get(name, (name, 999))
+
+        all_connections = shop_model["connections"]
+        all_junctions = shop_model["model"].get("junction", {})
+        all_tunnels = shop_model["model"].get("tunnel", {})
+        inlet_reservoir_name, connection_losses = self._plant_to_inlet_reservoir_with_losses(
+            name, all_connections, all_junctions, all_tunnels, set(reservoir_by_name.keys())
+        )
+
+        plant_generators: dict[str, GeneratorWrite] = {}
+        for connection in all_connections:
+            if (
+                connection.get("from_type") == "plant"
+                and connection["from"] == name
+                and (gen := generator_by_name.get(connection["to"]))
+            ):
+                plant_generators[gen.name] = gen
+            elif (
+                connection.get("to_type") == "plant"
+                and connection["to"] == name
+                and (gen := generator_by_name.get(connection["from"]))
+            ):
+                plant_generators[gen.name] = gen
```

Comment:
**issue:** This will overwrite any duplicates and use the last item processed if there are duplicates without informing the user that duplicates were encountered

Reply:
We do not care whether there is a connection to or from the generator. Thus, if both are present we are ok with overwriting it.